### PR TITLE
AP_RangeFinder: fixed LeddarOne busy wait

### PR DESCRIPTION
**Pull request by ShingoMatsuura**
Fixed LeddarOne busy wait. remove waiting parts for uart->available(). 

I tested flying with LeddarOne.
I uploaded flight log and CTUN/SAlt Graph image.

[Flight log file](https://dl.dropboxusercontent.com/u/2586545/LeddarOne/2016-11-03%2013-56-22.bin)

[CTUN/SAlt Graph](https://dl.dropboxusercontent.com/u/2586545/LeddarOne/CTUN-Alt-SAlt.emf)

### CODE REVIEW

File: libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp [hunk 1]
```diff
@@ -130,33 +161,35 @@ bool AP_RangeFinder_LeddarOne::CRC16(uint8_t *aBuffer, uint8_t aLength, bool aCh
...
 
     // Modbus read input register (function code 0x04)
-    data_buffer[0] = LEDDARONE_DEFAULT_ADDRESS;
-    data_buffer[1] = 0x04;
-    data_buffer[2] = 0;
-    data_buffer[3] = 20;
-    data_buffer[4] = 0;
-    data_buffer[5] = 10;
+    // send_buffer[3] = 20: Address of first register to read
+    // send_buffer[5] = 10: The number of consecutive registers to read
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp]: these comments belong down on the corresponding line

File: libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp [hunk 2]
```diff
@@ -130,33 +161,35 @@ bool AP_RangeFinder_LeddarOne::CRC16(uint8_t *aBuffer, uint8_t aLength, bool aCh
  */
 LeddarOne_Status AP_RangeFinder_LeddarOne::send_request(void)
 {
-    uint8_t data_buffer[10] = {0};
-    uint8_t i = 0;
+    uint8_t send_buffer[10] = {0};
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp]: the for loop goes 8 times but the index is size 10? There is 6 for data and 2 for CRC. Shouldn't this be [8]?

File: libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp [hunk 3]
```diff
@@ -130,33 +161,35 @@ bool AP_RangeFinder_LeddarOne::CRC16(uint8_t *aBuffer, uint8_t aLength, bool aCh
...
+    uint8_t index = 0;
 
     uint32_t nbytes = uart->available();
 
     // clear buffer
     while (nbytes-- > 0) {
         uart->read();
-        if (++i > 250) {
+        if (++index > 250) {
             return LEDDARONE_ERR_SERIAL_PORT;
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp]: I don't understand why reading a lot is an error? In your read state you should just empty the input buffer and accept as many samples as there are - and detect the correct framing from there.

File: libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.h [hunk 4]
```diff
@@ -2,16 +2,17 @@
...
 #define LEDDARONE_DETECTIONS_MAX 3
 
 // default slave address
 #define LEDDARONE_DEFAULT_ADDRESS 0x01
+#define LEDDARONE_DATA_INDEX_OFFSET 11
 
 // LeddarOne status
 enum LeddarOne_Status {
     LEDDARONE_OK = 0,
+	LEDDARONE_READING_BUFFER = 1,
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.h]: whitespace

File: libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp [hunk 5]
```diff
@@ -48,36 +48,67 @@ bool AP_RangeFinder_LeddarOne::detect(RangeFinder &_ranger, uint8_t instance, AP
...
+            if (leddarone_status == LEDDARONE_OK) {
+                reading_cm = sum_distance / number_detections;
+
+                // reset mod_bus status to read new buffer
+                modbus_status = LEDDARONE_MODBUS_PRE_SEND_REQUEST;
+
+                return true;
+            }
+            // keep reading next buffer
+            else if (leddarone_status == LEDDARONE_READING_BUFFER) {
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp]: wouldn't this be simplified by using != READING_BUFFER and moving the logic up into this line? An else break is more logical

File: libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp (hunk 1 above)
Comment (ShingoMatsuura) [inline libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp]: OK. Im going to change.

File: libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp (hunk 2 above)
Comment (ShingoMatsuura) [inline libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp]: Yes. I agree. thanks

File: libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.h (hunk 4 above)
Comment (ShingoMatsuura) [inline libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.h]: Thank. Im going to fix. LeddarOne_ModbusStatus has whitespace too.

File: libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp (hunk 5 above)
Comment (ShingoMatsuura) [inline libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp]: Yes, I agree. Im going to fix.

File: libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp [hunk 6]
```diff
@@ -168,40 +201,38 @@ LeddarOne_Status AP_RangeFinder_LeddarOne::send_request(void)
...
-            len += nbytes;
+            read_buffer[index] = uart->read();
+        }
+
+        read_len += nbytes;
+
+        if (read_len < 25) {
+            return LEDDARONE_READING_BUFFER;
         }
     }
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp]: would be nice if the protocol packet was described somewhere, either here or top of function. Here's what I've gathered so far in an expected packet:
read_buffer[0] == ??? Sync?
read_buffer[1] == 0x04 = ??? Sync?? device type? mfgr?
read_buffer[2] == ???
read_buffer[3] == ???
...
read_buffer[10] == number of detections/reflections
read_buffer[11] == LEDDARONE_DATA_INDEX_OFFSET == first data index
...
read_buffer[12] == last index of data
read_buffer[23] = CRC.Low???
read_buffer[24] = CRC.High???

File: libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp (hunk 3 above)
Comment (ShingoMatsuura) [inline libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp]: I am not sure, Arduino LeddarOne sample clears before sending request. I measured how log send_request functions takes. It has finished under 0ms so I think available buffer might be zero. I removed the part and test.

File: libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp (hunk 6 above)
Comment (ShingoMatsuura) [inline libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp]: Yes, I agree. Im going recheck LeddarOne document. Thank you for gathering!

File: libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp [hunk 7]
```diff
@@ -48,36 +48,67 @@ bool AP_RangeFinder_LeddarOne::detect(RangeFinder &_ranger, uint8_t instance, AP
...
-    uint32_t start_ms = AP_HAL::millis();
-    while (!uart->available()) {
-        // wait up to 200ms
-        if (AP_HAL::millis() - start_ms > 200) {
-            return false;
-        }
-    }
+        case LEDDARONE_MODBUS_PRE_SEND_REQUEST:
+            // clear buffer and buffer_len
+            memset(read_buffer, 0, sizeof(read_buffer));
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp]: if you have a lose connection then your state machine will resend and continually clear this buffer. There should be an additional state of "init" which is set on a complete packet but skipped if we get no data. To make it faster, that init state will set to the next state and then fall through to that next state of sending but a repeat send on timeout would skip the init

File: libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp (hunk 3 above)
Comment (ShingoMatsuura) [inline libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp]: OK. Thanks.

**Comment by OXINARF** [comment]
@magicrub I've assigned this to Randy and Tridge as they have gone through the driver and know what needed to change. But if you feel it is good to go in, feel free :smiley:

**Comment by muramura** [comment]
Coding style of the switch statement is different.
I think it is better to meet the coding style.

**Comment by muramura** [comment]
Magic number has been is used.
I think it is better to define the meaning of the value.

**Comment by ShingoMatsuura** [comment]
Thanks review @muramura - san. Yon mentioned about removing "break" on LEDDAR_MODBUS_INIT case?

Yes, I think so but It's better fall through to SEND_REQUEST to do send_request faster. It is rare case...

**Comment by ShingoMatsuura** [comment]
Thanks @muramura -san. Im going to change magic number to define.

**Comment by muramura** [comment]
Coding style of the switch statement is defined like this.
The start of the switch and the case is the same column.
===
switch (condition) {
case foo_cond:
    foo();
    break;
case bar_cond:
    bar();
    break;
}

**Comment by ShingoMatsuura** [comment]
Thanks @muramura - san. I saw stye-guide link and removed indent of case statements.

File: libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp [hunk 8]
```diff
@@ -48,36 +48,64 @@ bool AP_RangeFinder_LeddarOne::detect(RangeFinder &_ranger, uint8_t instance, AP
...
+        modbus_status = LEDDARONE_MODBUS_SENT_REQUEST;
+        last_sending_request_ms = AP_HAL::millis();
+        break;
+
+    case LEDDARONE_MODBUS_SENT_REQUEST:
+        if (uart->available()) {
+            // change mod_bus status to read available buffer
+            modbus_status = LEDDARONE_MODBUS_AVAILABLE;
+        } else {
+            if (AP_HAL::millis() - last_sending_request_ms > 200) {
```
Comment (muramura) [inline libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp]: This 200ms and Will value that came out from where is.
It is also used in other driver.
For this 200ms, I think that the case of the driver common time-out value, it is better defined in the common header.

File: libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp [hunk 9]
```diff
@@ -48,36 +48,64 @@ bool AP_RangeFinder_LeddarOne::detect(RangeFinder &_ranger, uint8_t instance, AP
...
-        // wait up to 200ms
-        if (AP_HAL::millis() - start_ms > 200) {
-            return false;
+    switch (modbus_status) {
+    case LEDDARONE_MODBUS_INIT:
+        // clear buffer and buffer_len
+        memset(read_buffer, 0, sizeof(read_buffer));
+        read_len = 0;
+        modbus_status = LEDDARONE_MODBUS_PRE_SEND_REQUEST;
+
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp]: Yes, which is why I said 
>Make sure to have a comment saying that it is intentional

Let's make sure to have that

File: libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp (hunk 8 above)
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp]: I agree

File: libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp (hunk 9 above)
Comment (ShingoMatsuura) [inline libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp]: Sorry I forgot adding comment. I added. Thanks comment.

File: libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.h [hunk 10]
```diff
@@ -20,6 +29,14 @@ enum LeddarOne_Status {
     LEDDARONE_ERR_NUMBER_DETECTIONS = -6
 };
 
+// LeddarOne Modbus status
```
Comment (magicrub) [inline libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.h]: This enumeration needs to be renamed to include the word: state

Comment (ShingoMatsuura) [inline libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.h]: Tkanks comment. I'm going to fix.

**Comment by magicrub** [comment]
Merged, thanks! @rmackay9 I tagged this for v3.4.2 but could go into .1 if possible.

**Comment by OXINARF** [comment]
@magicrub .1 is already out so not possible :smile:

**Comment by ShingoMatsuura** [comment]
Thanks merge and reviewing @magicrub - san.

**Pull request merged**